backend/daemon/api/entities/v1alpha/entities.go

```diff
@@ -400,6 +398,160 @@ func (api *Server) SearchEntities(ctx context.Context, in *entities.SearchEntiti
+					WHERE multihash = :hash
+				`)
+				_, err = hypersql.BlobsDelete(conn, c.Hash())
+				if err != nil {
+					_, err = &emptypb.Empty{}, api.blobs.Query(ctx, func(conn *sqlite.Conn) error {
+						return sqlitex.Exec(conn, qEmptyComments(), func(_ *sqlite.Stmt) error {
+							return nil
+						}, c.Hash())
+					})
+					if err != nil {
+						return err
+					}
+				}
+				if cmt.RepliedComment.String() != "" {
+					_, err = hypersql.BlobsDelete(conn, cmt.RepliedComment.Hash())
+					if err != nil {
+						_, err = &emptypb.Empty{}, api.blobs.Query(ctx, func(conn *sqlite.Conn) error {
+							return sqlitex.Exec(conn, qEmptyComments(), func(_ *sqlite.Stmt) error {
+								return nil
+							}, cmt.RepliedComment.Hash())
+						})
+						if err != nil {
+							return err
+						}
+					}
+				}
+
+				return nil
+			}
+			return nil
+		})
+	})
+	err = api.blobs.DeleteEntity(ctx, eid, meta)
+	if err != nil {
+		if errors.Is(err, hyper.ErrEntityNotFound) {
+			return nil, err
+		}
+
+		var qEmptyBlobs = dqb.Str(`
+		UPDATE blobs
+	    SET data = NULL, size = -1
+		WHERE id in (
+			SELECT blob_id from structural_blobs_view where resource = :eid
+		)
+		`)
+
+		_, err = &emptypb.Empty{}, api.blobs.Query(ctx, func(conn *sqlite.Conn) error {
+			return sqlitex.Exec(conn, qEmptyBlobs(), func(_ *sqlite.Stmt) error {
+				return nil
+			}, in.Id)
+		})
+		if err != nil {
+			return &emptypb.Empty{}, err
+		}
+		var qDeleteStructuralBlobs = dqb.Str(`
+		DELETE from structural_blobs
+		WHERE id in (
+			SELECT blob_id from structural_blobs_view where resource = :eid
+		)
+		`)
+		_, err = &emptypb.Empty{}, api.blobs.Query(ctx, func(conn *sqlite.Conn) error {
+			return sqlitex.Exec(conn, qDeleteStructuralBlobs(), func(_ *sqlite.Stmt) error {
+				return nil
+			}, in.Id)
+		})
+		if err != nil {
+			return &emptypb.Empty{}, err
+		}
+	}
+	_, err = &emptypb.Empty{}, api.blobs.Query(ctx, func(conn *sqlite.Conn) error {
+		return sqlitex.WithTx(conn, func() error {
+			res, err := hypersql.EntitiesInsertRemovedRecord(conn, eid.String(), in.Reason, meta)
+			if err != nil {
+				return err
+			}
+			if res.ResourceEID != eid.String() {
+				return fmt.Errorf("%w: %s", hyper.ErrEntityNotFound, eid)
+			}
+
+			return nil
+		})
+	})
+	return &emptypb.Empty{}, err
+}
+
+// RestoreEntity implements the corresponding gRPC method.
+func (api *Server) RestoreEntity(ctx context.Context, in *entities.RestoreEntityRequest) (*emptypb.Empty, error) {
+	if in.Id == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "must specify entity ID to restore")
+	}
+
+	eid := hyper.EntityID(in.Id)
+
+	return &emptypb.Empty{}, api.blobs.Query(ctx, func(conn *sqlite.Conn) error {
+		return hypersql.EntitiesDeleteRemovedRecord(conn, eid.String())
+	})
+}
+
+// ListDeletedEntities implements the corresponding gRPC method.
+func (api *Server) ListDeletedEntities(ctx context.Context, in *entities.ListDeletedEntitiesRequest) (*entities.ListDeletedEntitiesResponse, error) {
```

Comment:
I did not include pagination here. Is it necessary @burdiyan ?

Reply:
As long as the client handles pagination as if it existed, I think we are fine.